network/p2p/pubsub.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package p2p
+
+import (
+	"context"
+	"time"
+
+	"github.com/algorand/go-algorand/config"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	pubsub_pb "github.com/libp2p/go-libp2p-pubsub/pb"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"golang.org/x/crypto/blake2b"
+)
+
+func init() {
+	// configure larger overlay parameters
+	pubsub.GossipSubD = 8
+	pubsub.GossipSubDscore = 6
+	pubsub.GossipSubDout = 3
+	pubsub.GossipSubDlo = 6
+	pubsub.GossipSubDhi = 12
+	pubsub.GossipSubDlazy = 12
+	pubsub.GossipSubDirectConnectInitialDelay = 30 * time.Second
+	pubsub.GossipSubIWantFollowupTime = 5 * time.Second
+	pubsub.GossipSubHistoryLength = 10
+	pubsub.GossipSubGossipFactor = 0.1
+}
+
+const (
+	gossipScoreThreshold             = -500
+	publishScoreThreshold            = -1000
+	graylistScoreThreshold           = -2500
+	acceptPXScoreThreshold           = 1000
+	opportunisticGraftScoreThreshold = 3.5
+)
+
+// TXTopicName defines a pubsub topic for TX messages
+const TXTopicName = "/algo/tx/0.1.0"
+
+func makePubSub(ctx context.Context, cfg config.Local, host host.Host) (*pubsub.PubSub, error) {
+	//defaultParams := pubsub.DefaultGossipSubParams()
+
+	options := []pubsub.Option{ // TODO
+		pubsub.WithPeerScore(&pubsub.PeerScoreParams{
+			DecayInterval: pubsub.DefaultDecayInterval,
+			DecayToZero:   pubsub.DefaultDecayToZero,
+
+			AppSpecificScore: func(p peer.ID) float64 { return 1000 }, // TODO
+
+			Topics: map[string]*pubsub.TopicScoreParams{
+				TXTopicName: {
+					TopicWeight: 0.1,
+
+					TimeInMeshWeight:  0.0002778, // ~1/3600
+					TimeInMeshQuantum: time.Second,
+					TimeInMeshCap:     1,
+
+					FirstMessageDeliveriesWeight: 0.5, // max value is 50
+					FirstMessageDeliveriesDecay:  pubsub.ScoreParameterDecay(10 * time.Minute),
+					FirstMessageDeliveriesCap:    100, // 100 messages in 10 minutes
+
+					// invalid messages decay after 1 hour
+					InvalidMessageDeliveriesWeight: -1000,
+					InvalidMessageDeliveriesDecay:  pubsub.ScoreParameterDecay(time.Hour),
+				},
+			},
```

Comment:
It feels like we should move topic-specific params/functions into their own module. I can envision something like `network/p2p/topics/tx.go`, `network/p2p/topics/agreement.go`, etc.